### PR TITLE
[web] Multiple fixes to text layout and line breaks

### PR DESCRIPTION
**Pull request by mdebbar**
- Modifications to the line breaking algorithm:
  - Allow break before/after parentheses and other punctuations when they are preceded/followed by whitespace.
  - Adjust the test fixtures accordingly.
- Fix line breaking when `maxEnd` coincides with the end of a whitespace sequence.
- Fix line breaking when there are consecutive non-breakable spans that don't fit in a single line.
- Fix some edge cases that cause text layout to incorrectly take prohibited line breaks.
- Avoid unnecessary recomputation of next line break.
- Fix some test helpers to show correct info on failures.
- Null safety improvements.

This required some screenshot updates: https://github.com/flutter/goldens/pull/225

Fixes https://github.com/flutter/flutter/issues/89586
Allows tests mentioned [here](https://github.com/flutter/flutter/issues/87696) to be unskipped.

### CODE REVIEW

File: lib/web_ui/test/text/layout_service_rich_test.dart [hunk 1]
```diff
@@ -223,4 +211,43 @@ Future<void> testMain() async {
...
+    expect(paragraph.height, 20.0);
+    expectLines(paragraph, <TestLine>[
+      // 1x small "A" + 19x big "A"
+      l('A' * 20, 0, 20, hardBreak: false, width: 1.0 + 190.0, height: 10.0),
+      // 1x big "A" + 1x small "A"
+      l('AA', 20, 22, hardBreak: true, width: 10.0 + 1.0, height: 10.0),
+    ]);
+  });
+
+  test('does not take prohibited line breaks', () {
```
Comment (yjbanov) [inline lib/web_ui/test/text/layout_service_rich_test.dart:236]: nit: s/take/make/ ?

File: lib/web_ui/test/text/line_breaker_test_helper.dart [hunk 2]
```diff
@@ -2,7 +2,6 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-
 // The following test cases contradict rule LB25, so we are replacing them
 // with correct expectations.
 const Map<String, String> _replacements = <String, String>{
```
Comment (yjbanov) [inline lib/web_ui/test/text/line_breaker_test_helper.dart:7]: Since these replacements are intentionally contradicting the Unicode rules, should we revise to make sure they don't contain `SPACE (SP) × [13.`? Otherwise, `_checkReplacement` will be contradicting our own contradictions :)

File: lib/web_ui/lib/src/engine/text/line_breaker.dart [hunk 3]
```diff
@@ -389,26 +404,31 @@ LineBreakResult nextLineBreak(String text, int index, {int? maxEnd}) {
...
     //       × IS
     //       × SY
-    if (curr == LineCharProperty.CL ||
-        curr == LineCharProperty.CP ||
-        curr == LineCharProperty.EX ||
-        curr == LineCharProperty.IS ||
-        curr == LineCharProperty.SY) {
+    //
+    // Modification: when these characters are preceded by a space, they
+    // shouldn't prevent a line break.
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/line_breaker.dart:409]: Correct, the above is a quote from unicode.org. What do you suggest to make it clear that it's a quote?

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 4]
```diff
@@ -178,6 +175,13 @@ class TextLayoutService {
             currentLine = currentLine.nextLine();
           } else {
             // Normal line break.
+            currentLine.revertToLastBreakOpportunity();
+            // If a revert had occurred in the line, we need to revert the span
+            // index accordingly.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:180]: Mention the special case that when no revert happened then `revertedToSpan` will simply be equal `span`.

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 5]
```diff
@@ -1182,6 +1204,18 @@ class LineBuilder {
...
 
+  /// Looks for the last break opportunity in the line and reverts the line to
+  /// that point.
+  ///
+  /// If the line already ends with a break opportunity, this method does
+  /// nothing.
+  void revertToLastBreakOpportunity() {
+    assert(isBreakable);
+    while (isEndProhibited) {
+      _popSegment();
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:1215]: Let's assert at the end of `_popSegment` that `isBreakable` is still true.

**Review by yjbanov — APPROVED** [review]
lgtm
